caffe/src/caffe/layers/molgrid_data_layer.cpp

```diff
@@ -688,122 +796,128 @@ void MolGridDataLayer<Dtype>::forward(const vector<Blob<Dtype>*>& bottom, const
 {
   bool hasaffinity = this->layer_param_.molgrid_data_param().has_affinity();
   bool hasrmsd = this->layer_param_.molgrid_data_param().has_rmsd();
-  Dtype *data = NULL;
+
+  Dtype *top_data = NULL;
   if(gpu)
-    data = top[0]->mutable_gpu_data();
+    top_data = top[0]->mutable_gpu_data();
   else
-    data = top[0]->mutable_cpu_data();
+    top_data = top[0]->mutable_cpu_data();
 
-  labels.resize(0);
-  affinities.resize(0);
-  rmsds.resize(0);
+  //clear batch labels
+  labels.clear();
+  affinities.clear();
+  rmsds.clear();
   unsigned batch_size = top_shape[0];
+
+  //percent of batch from first data source
+  float pct_data = 1.0;
+  if (two_data_sources)
+    pct_data = data_ratio/(data_ratio+1);
```

Comment:
data_ratio is the relative amount of data from source1 vs. source2, so 2.0 would mean train on twice as much source1 as source2. I convert this to pct_data, the percent of each batch that should come from source1 (2.0 becomes 0.67) and the rest from source2. data_ratio seems more user-friendly so it's the prototxt parameter, but it's easier to fill in the batches with data by comparing an index to pct_data*batch_size to determine which source to get an example from.